nightfall-client/src/services/transform.mjs

```diff
@@ -0,0 +1,234 @@
+/* eslint-disable no-await-in-loop */
+
+import config from 'config';
+import constants from '@polygon-nightfall/common-files/constants/index.mjs';
+import logger from '@polygon-nightfall/common-files/utils/logger.mjs';
+import { randValueLT } from '@polygon-nightfall/common-files/utils/crypto/crypto-random.mjs';
+import { waitForContract } from '@polygon-nightfall/common-files/utils/contract.mjs';
+import {
+  getCircuitHash,
+  generateProof,
+} from '@polygon-nightfall/common-files/utils/worker-calls.mjs';
+import utils from '@polygon-nightfall/common-files/utils/crypto/merkle-tree/utils.mjs';
+import gen from 'general-number';
+import { Commitment, Transaction } from '../classes/index.mjs';
+import { clearPending } from './commitment-storage.mjs';
+import { getCommitmentInfo } from '../utils/getCommitmentInfo.mjs';
+import { computeCircuitInputs } from '../utils/computeCircuitInputs.mjs';
+import { submitTransaction } from '../utils/submitTransaction.mjs';
+import { ZkpKeys } from './keys.mjs';
+
+const { VK_IDS } = config;
+const { SHIELD_CONTRACT_NAME, BN128_GROUP_ORDER } = constants;
+const { generalise } = gen;
+
+async function transform(transformParams) {
+  logger.info('Creating a transform transaction');
+
+  const { inputTokens, outputTokens, ...items } = transformParams;
+  const { rootKey, compressedZkpPublicKey, fee } = generalise(items);
+
+  const shieldContractInstance = await waitForContract(SHIELD_CONTRACT_NAME);
+  const maticAddress = generalise(
+    (await shieldContractInstance.methods.getMaticAddress().call()).toLowerCase(),
+  );
+
+  const circuitHash = await getCircuitHash('transform');
+
+  // the last two nullifiers are reserverd for covering the fee
```

Comment:
im not a big fan of this logic.

Reserving nullifiers and commitments to be used as part of fees leads to unnecessary rigidity, reduces anonymity and means that on-chain transactions will be padded such that non-zero hashes are between 0-hashes.

Reply:
If the transaction takes up to N nullifiers and C commitments, then we should allow for:

- Up to (N-1) nullifiers related to the fee and up to 1 commitment related to the change or;
- Up to  N nullifiers related to the token transformation and up to C commitments related to the token change.

in my mind the number of commitments, C, should be at least N + 1 (maybe even N + 2) to account for a partial transformation of input commitments and the fee.

Reply:
> in my mind the number of commitments, C, should be at least N + 1 (maybe even N + 2) to account for a partial transformation of input commitments and the fee.
I agree with this, right now this is set to C=5 N=6 but N=7 seems reasonable to me as well.

My reason for reserving the last two commitments for the fee were to match the burn and tokenise transaction. 
Another concern I had was that checking if a commitment is a fee or not on the circuit level would increase the number of constraints.

Depending on your thoughts on this I'm happy to change it to be completely agnostic to the order of commitments / nullifiers.